File: Leetcode_All_Solved/162.cpp
class Solution {
private:
    int binary_upwards(const vector<int>& nums, int low, int high){
        if(low == high)return low;
        int mid = low + (high-low)/2;
        if(nums[mid] < nums[mid+1])//mid+1 <= high_max
            return binary_upwards(nums, mid+1, high);
        else return binary_upwards(nums, low, mid);
    }
public:
    int findPeakElement(vector<int>& nums) {
        //NOTE: nums[i] != nums[i+1] for all valid i
        return binary_upwards(nums, 0, nums.size()-1);
    }
};

File: Leetcode_All_Solved/162.py
# O(logn)
class Solution:
    def findPeakElement(self, nums: List[int]) -> int:
        def binary_upwards(nums: list[int], low: int, high: int) -> int:
            if(low == high):
                return low
            
            mid = (high+low)//2
            # since nums[i] != nums[i+1]
            # mid+1 < n, already ruled out by if low == high
            if(nums[mid] < nums[mid+1]):
                return binary_upwards(nums, mid+1, high)
            else:
                return binary_upwards(nums, low, mid)
        
        return binary_upwards(nums, 0, len(nums)-1)
